internal/appsec/limiter_test.go

```diff
@@ -0,0 +1,251 @@
+		l := NewTokenTicker(100, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		for i := 0; i < 100; i++ {
+			require.True(t, l.Allow(),
+				"Burst call %d to limiter.Allow() should return True with 100 initial tokens", i)
+			startTime = startTime.Add(50 * time.Microsecond)
+			ticksChan <- startTime
+		}
+		require.False(t, l.Allow(),
+			"Burst call 101 to limiter.Allow() should return False with 100 initial tokens")
+	})
+
+	t.Run("bucket-refill-short", func(t *testing.T) {
+		l := NewTokenTicker(100, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+
+		for i := 0; i < 1000; i++ {
+			startTime.Add(time.Millisecond)
+			ticksChan <- startTime
+			require.Equalf(t, int64(100), atomic.LoadInt64(&l.tokens), "Bucket should have exactly 100 tokens")
+		}
+	})
+
+	t.Run("bucket-refill-long", func(t *testing.T) {
+		l := NewTokenTicker(100, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+
+		for i := 0; i < 1000; i++ {
+			startTime = startTime.Add(3 * time.Second)
+			ticksChan <- startTime
+		}
+		require.Equalf(t, int64(100), atomic.LoadInt64(&l.tokens), "Bucket should have exactly 100 tokens")
+	})
+}
+
+func TestLimiter(t *testing.T) {
+	t.Run("allow-after-stop", func(t *testing.T) {
+		l := NewTokenTicker(3, 3)
+		l.Start()
+		require.True(t, l.Allow())
+		l.Stop()
+		// The limiter keeps allowing until there's no more tokens
+		require.True(t, l.Allow())
+		require.True(t, l.Allow())
+		require.False(t, l.Allow())
+	})
+
+	t.Run("concurrency", func(t *testing.T) {
+		//Tests the limiter's ability to sample the traces when subjected to a continuous flow of requests
+		//Each goroutine will continuously call the WAF and the rate limiter for 1 second
+		for nbUsers := 1; nbUsers <= 1000; nbUsers *= 10 {
+			t.Run(fmt.Sprintf("continuous-requests-%d-users", nbUsers), func(t *testing.T) {
+				var startBarrier, stopBarrier sync.WaitGroup
+				// Create a start barrier to synchronize every goroutine's launch and
+				// increase the chances of parallel accesses
+				startBarrier.Add(1)
+				// Create a stopBarrier to signal when all user goroutines are done.
+				stopBarrier.Add(nbUsers)
+				skipped := int64(0)
+				kept := int64(0)
+				l := NewTokenTicker(0, 100)
+
+				for n := 0; n < nbUsers; n++ {
+					go func(l Limiter, kept *int64, skipped *int64) {
+						startBarrier.Wait()      // Sync the starts of the goroutines
+						defer stopBarrier.Done() // Signal we are done when returning
+
+						for tStart := time.Now(); time.Since(tStart) < 1*time.Second; {
+							if !l.Allow() {
+								atomic.AddInt64(skipped, 1)
+							} else {
+								atomic.AddInt64(kept, 1)
+							}
+						}
+					}(l, &kept, &skipped)
+				}
+
+				l.Start()
+				defer l.Stop()
+				start := time.Now()
+				startBarrier.Done() // Unblock the user goroutines
+				stopBarrier.Wait()  // Wait for the user goroutines to be done
+				duration := time.Since(start).Seconds()
+				//Limiter started with 1 token, expecting a margin of error of 1
```

Comment:
The limiter was created with 0 tokens: https://github.com/DataDog/dd-trace-go/pull/1131/files#diff-dfdcd83a96ec89da66b656478722ea9d3f4c92c4daa42ccfeb485793e6c283c8R151 but the comment suggests 1.

Reply:
Nice catch. This comment can be removed.